src/server/pkg/storage/chunk/transform.go

```diff
@@ -0,0 +1,201 @@
+	switch algo {
+	case CompressionAlgo_NONE:
+		copy(dst, src)
+		return CompressionAlgo_NONE, len(src), nil
+	case CompressionAlgo_GZIP_BEST_SPEED:
+		lw := newLimitWriter(dst)
+		err := func() error {
+			gw, err := gzip.NewWriterLevel(lw, gzip.BestSpeed)
+			if err != nil {
+				return err
+			}
+			_, err = gw.Write(src)
+			if err != nil {
+				return err
+			}
+			return gw.Close()
+		}()
+		if err == io.ErrShortWrite {
+			return compress(CompressionAlgo_NONE, dst, src)
+		}
+		return CompressionAlgo_GZIP_BEST_SPEED, lw.pos, err
+	default:
+		return 0, 0, errors.Errorf("unrecognized compression: %v", algo)
+	}
+}
+
+func decompress(algo CompressionAlgo, r io.Reader) (io.Reader, error) {
+	switch algo {
+	case CompressionAlgo_NONE:
+		return r, nil
+	case CompressionAlgo_GZIP_BEST_SPEED:
+		gr, err := gzip.NewReader(r)
+		if err != nil {
+			return nil, err
+		}
+		return gr, nil
+	default:
+		return nil, errors.Errorf("unrecognized compression: %v", algo)
+	}
+}
+
+type limitWriter struct {
+	buf []byte
+	pos int
+}
+
+func newLimitWriter(buf []byte) *limitWriter {
+	return &limitWriter{buf: buf}
+}
+
+func (w *limitWriter) Write(p []byte) (int, error) {
+	if w.pos >= len(w.buf) {
+		return 0, io.ErrShortWrite
+	}
+	n := copy(w.buf[w.pos:], p)
+	if n < len(p) {
+		return n, io.ErrShortWrite
+	}
+	w.pos += n
+	return n, nil
+}
+
+// encrypt generates a key using secret and src, and encrypts src, writing the output to dst
+func encrypt(secret []byte, dst, src []byte) (dek []byte) {
+	dek = deriveKey(secret, src)
+	cryptoXOR(dek[:32], dst, src)
+	return dek
+}
+
+// decrypt returns an io.Reader containing r decrypted using dek
+func decrypt(dek []byte, r io.Reader) (io.Reader, error) {
+	if len(dek) != 32 {
+		return nil, errors.Errorf("data encryption key is wrong length")
+	}
+	nonce := [chacha20.NonceSize]byte{}
+	ciph, err := chacha20.NewUnauthenticatedCipher(dek, nonce[:])
+	if err != nil {
+		return nil, err
+	}
+	return cipher.StreamReader{S: ciph, R: r}, nil
+}
+
+// deriveKey returns Hash(secret + Hash(ptext))
+func deriveKey(secret, ptext []byte) []byte {
+	var x []byte
+	x = append(x, secret...)
+	x = append(x, Hash(ptext)...)
+	return Hash(x)[:32]
+}
+
+// cryptoXOR setups up a stream cipher using key, and writes (src XOR keystream) to dst
+func cryptoXOR(key, dst, src []byte) {
+	nonce := [chacha20.NonceSize]byte{}
+	ciph, err := chacha20.NewUnauthenticatedCipher(key, nonce[:])
+	if err != nil {
+		panic(err) // this only happens if you pass in an invalid key/nonce size, which we shouldn't do
+	}
+	ciph.XORKeyStream(dst, src)
+}
+
+func (r *Ref) Key() string {
+	data, err := r.Marshal()
+	if err != nil {
+		panic(err)
+	}
+	return hex.EncodeToString(Hash(data)[:16])
+}
+
+func getFromCache(ctx context.Context, objs obj.Client, ref *Ref, w io.Writer) error {
+	rc, err := objs.Reader(ctx, ref.Key(), 0, 0)
+	if err != nil {
+		return err
+	}
+	defer rc.Close()
```

Comment:
error

Reply:
The object client stuff is now taken care of in `storage/kv/obj_adapter.go`